src/forms/TimeInput.jsx

```diff
@@ -19,8 +50,91 @@ class TimeInput extends React.Component {
 	* @description called when the input changes, in turn calls the onChange
 	* 	handler prop, if there is one provided (eg supplied by redux-form or DateTimePicker)
 	*/
-	onChange(e) {
-		this.props.onChange && this.props.onChange(e);
+	onChange(partialState) {
+		if (this.props.onChange) {
+			const stateValues = {
+				...this.state,
+				...partialState
+			};
+
+			const v = `${formatDigits(stateValues.hours) % (this.props.is24Hr ? 24 : 12)}:${formatDigits(stateValues.minutes)}`;
+			this.props.onChange(v);
+		}
+	}
+
+	/**
+	* @function onTimeInputChange
+	* @param e Event Object
+	* @description called when the browser-native time input changes, in turn calls the onChange
+	* 	handler prop, if there is one provided (eg supplied by redux-form or DateTimePicker)
+	*/
+	onTimeInputChange(e) {
+		const { value } = e.target;
+		this.setState(() => ({ value: value }));
+
+		this.props.onChange && this.props.onChange(value);
+	}
+
+	/**
+	* @function onNumberChange
+	* @param e Event Object
+	* @description called when either the hour or minute input changes,
+	* 	in turn changes the state and updates the value of the inputs
+	*/
+	onNumberChange(e) {
+		const { value, id } = e.target;
+
+		this.setState(() => ({ [id]: value }));
+	}
+	/**
+	* @function onMeridianChange
+	* @param e Event Object
+	* @description called when the meridian <select> input changes,
+	* 	in turn changes the state and updates the value of the <select>
+	*/
+	onMeridianChange(e) {
+		const { value } = e.target;
+
+		this.setState(() => ({ meridian: value }));
+	}
+
+	/**
+	* @function onBlur
+	* @param e Event Object
+	* @description called when either the hour or minute input loses focus,
+	* 	and ensures the value entered is not out of the min/max range
+	*/
+	onBlur(e) {
+		const { value, min, max, id } = e.target;
+
+		if (max || min) {
+			const constrainedVal = Math.max(Math.min(value, max), min);
+			if (constrainedVal == value) {
```

Comment:
is this the only place to constrain the value? you can factor it out into just its own mothod that returns a value
```constrainValue(value, min, max) ...```

if anything - I think this might be more leg:
```
let newValue = value;
if (max || min) { // constrain values
    newValue = Math.max(Math.min(value, max), min);
}
const newState = { [id]: formatDigits(newValuel) };
this.setState(() => (newState)); // try doing onChange stuff in this setstate callback
this.onChange(newState);
```

Reply:
I really like the idea to break `constrainValue` out into a function.

In terms of signature, I'd recommend

```js
constrainValue({ min=0, max=Infinity }, value)
```

that will allow you to have an _optional_ `min` and `max` param, and then you can pass in `value` as a second argument to make it easier to curry the function like

```js
const constrainTo100 = constrainValue.bind(null, { max:100 });
const contrainedValue = constrainTo100(value);
```

(not that you'd ever need to do that kind of currying, but it helps explain why having `value` as the _last_ argument is better than having it as the _first_ arg.